File: day08/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math/bits"
	"os"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

type (
	Node struct {
		ID    string
		Left  string
		Right string
		IsEnd bool
	}

	NodeVisit struct {
		ID        string
		Count     int
		Cycle     int
		Rem       int
		Candidate bool
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var steps []byte
	nodes := map[string]Node{}
	var starts []string

	first := true

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if first {
			first = false
			steps = []byte(scanner.Text())
			continue
		}
		line := scanner.Text()
		if line == "" {
			continue
		}
		lhs, rhs, ok := strings.Cut(line, " = ")
		if !ok {
			log.Fatalln("Invalid line")
		}
		a, b, ok := strings.Cut(strings.Trim(rhs, "()"), ", ")
		if !ok {
			log.Fatalln("Invalid line")
		}
		nodes[lhs] = Node{
			ID:    lhs,
			Left:  a,
			Right: b,
			IsEnd: strings.HasSuffix(lhs, "Z"),
		}
		if strings.HasSuffix(lhs, "A") {
			starts = append(starts, lhs)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	start, ok := nodes["AAA"]
	if !ok {
		log.Fatalln("No start node", "AAA")
	}
	end := "ZZZ"
	count := 0
	for start.ID != end {
		instr := steps[count%len(steps)]
		next := start.Left
		if instr == 'R' {
			next = start.Right
		}
		var ok bool
		start, ok = nodes[next]
		if !ok {
			log.Fatalln("Next node not found")
		}
		count++
	}
	fmt.Println("Part 1:", count)

	startNodes := make([]Node, 0, len(starts))
	for _, i := range starts {
		node, ok := nodes[i]
		if !ok {
			log.Fatalln("No start node", i)
		}
		startNodes = append(startNodes, node)
	}
	count = 0
	revisitNodes := make([]NodeVisit, len(startNodes))
	totalRevisits := 0
	totalStarts := len(startNodes)
	for totalRevisits < totalStarts {
		instr := steps[count%len(steps)]
		count++
		for i := range startNodes {
			next := startNodes[i].Left
			if instr == 'R' {
				next = startNodes[i].Right
			}
			node, ok := nodes[next]
			if !ok {
				log.Fatalln("Next node not found")
			}
			startNodes[i] = node
			if node.IsEnd {
				if revisitNodes[i].ID == "" {
					revisitNodes[i] = NodeVisit{
						ID:    node.ID,
						Count: count,
					}
				} else if !revisitNodes[i].Candidate {
					if node.ID != revisitNodes[i].ID {
						log.Fatalln("Multiple terminals for cycle")
					}
					cycle := count - revisitNodes[i].Count
					revisitNodes[i] = NodeVisit{
						ID:        revisitNodes[i].ID,
						Count:     count,
						Cycle:     cycle,
						Rem:       count % cycle,
						Candidate: true,
					}
					totalRevisits++
				} else {
					if node.ID != revisitNodes[i].ID {
						log.Fatalln("Multiple terminals for cycle")
					}
					if cycle := count - revisitNodes[i].Count; cycle != revisitNodes[i].Cycle {
						log.Fatalln("Multiple cycle lengths")
					}
					revisitNodes[i] = NodeVisit{
						ID:        revisitNodes[i].ID,
						Count:     count,
						Cycle:     revisitNodes[i].Cycle,
						Rem:       revisitNodes[i].Rem,
						Candidate: true,
					}
				}
			}
		}
	}
	a := revisitNodes[0].Rem
	m := revisitNodes[0].Cycle
	for _, i := range revisitNodes[1:] {
		var ok bool
		a, m, ok = crt(a, m, i.Rem, i.Cycle)
		if !ok {
			log.Fatalln("Unsolvable constraints")
		}
	}
	if a <= 0 {
		a += m
	}
	fmt.Println("Part 2:", a)
}

func crt(a1, m1, a2, m2 int) (int, int, bool) {
	g, p, q := extGCD(m1, m2)
	if a1%g != a2%g {
		return 0, 0, false
	}
	m1g := m1 / g
	m2g := m2 / g
	lcm := m1g * m2
	// a1 * m2/g * q + a2 * m1/g * p (mod lcm)
	x := (mulmod(mulmod(a1, m2g, lcm), q, lcm) + mulmod(mulmod(a2, m1g, lcm), p, lcm)) % lcm
	if x < 0 {
		x += lcm
	}
	return x, lcm, true
}

func extGCD(a, b int) (int, int, int) {
	x2 := 1
	x1 := 0
	y2 := 0
	y1 := 1
	// a should be larger than b
	flip := false
	if a < b {
		a, b = b, a
		flip = true
	}
	for b > 0 {
		q := a / b
		a, b = b, a%b
		x2, x1 = x1, x2-q*x1
		y2, y1 = y1, y2-q*y1
	}
	if flip {
		x2, y2 = y2, x2
	}
	return a, x2, y2
}

func mulmod(a, b, m int) int {
	sign := 1
	if a < 0 {
		a = -a
		sign *= -1
	}
	if b < 0 {
		b = -b
		sign *= -1
	}
	a = a % m
	b = b % m
	hi, lo := bits.Mul(uint(a), uint(b))
	return sign * int(bits.Rem(hi, lo, uint(m)))
}

File: day08/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

struct Node {
    id: String,
    left: String,
    right: String,
    is_end: bool,
}

struct NodeVisit<'a> {
    id: &'a str,
    count: usize,
    cycle: usize,
    candidate: bool,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut steps = Vec::new();
    let mut nodes = HashMap::new();
    let mut starts = Vec::new();
    for line in reader.lines() {
        let line = line?;
        if steps.len() == 0 {
            steps = line.into_bytes();
            continue;
        }
        if line == "" {
            continue;
        }
        let (lhs, rhs) = line.split_once(" = ").ok_or("Invalid line")?;
        let discard_chars: &[_] = &['(', ')'];
        let (a, b) = rhs
            .trim_matches(discard_chars)
            .split_once(", ")
            .ok_or("Invalid line")?;
        nodes.insert(
            lhs.to_owned(),
            Node {
                id: lhs.to_owned(),
                left: a.to_owned(),
                right: b.to_owned(),
                is_end: lhs.ends_with("Z"),
            },
        );
        if lhs.ends_with("A") {
            starts.push(lhs.to_owned());
        }
    }

    let mut start = nodes.get("AAA").ok_or("No start node")?;
    let end = "ZZZ";
    let mut count = 0;
    while start.id != end {
        let next = if steps[count % steps.len()] == b'L' {
            &start.left
        } else {
            &start.right
        };
        start = nodes.get(next).ok_or("Next node not found")?;
        count += 1;
    }

    println!("Part 1: {}", count);

    let mut start_nodes = starts
        .into_iter()
        .map(|v| nodes.get(&v).ok_or("Start node not found"))
        .collect::<Result<Vec<_>, _>>()?;
    let total_starts = start_nodes.len();
    let mut revisits = (0..total_starts)
        .map(|_| NodeVisit {
            id: "",
            count: 0,
            cycle: 0,
            candidate: false,
        })
        .collect::<Vec<_>>();
    let mut total_revisits = 0;
    let mut count = 0;
    while total_revisits < total_starts {
        let instr = steps[count % steps.len()];
        count += 1;
        for (i, v) in start_nodes.iter_mut().enumerate() {
            let next = if instr == b'L' { &v.left } else { &v.right };
            *v = nodes.get(next).ok_or("Next node not found")?;
            if v.is_end {
                let visit = revisits.get_mut(i).ok_or("Revisit does not exist")?;
                if visit.id == "" {
                    *visit = NodeVisit {
                        id: &v.id,
                        count,
                        cycle: 0,
                        candidate: false,
                    };
                } else if !visit.candidate {
                    if v.id != visit.id {
                        return Err("Multiple terminals for cycle".into());
                    }
                    let cycle = count - visit.count;
                    let rem = count % cycle;
                    if rem != 0 {
                        return Err("Remainder not zero".into());
                    }
                    *visit = NodeVisit {
                        id: visit.id,
                        count,
                        cycle,
                        candidate: true,
                    };
                    total_revisits += 1;
                } else {
                    if v.id != visit.id {
                        return Err("Multiple terminals for cycle".into());
                    }
                    let cycle = count - visit.count;
                    if cycle != visit.cycle {
                        return Err("Multiple cycle lengths".into());
                    }
                    visit.count = count;
                }
            }
        }
    }

    println!(
        "Part 2: {}",
        revisits.into_iter().fold(1, |acc, v| lcm(acc, v.cycle))
    );

    Ok(())
}

fn lcm(a: usize, b: usize) -> usize {
    a * b / gcd(a, b)
}

fn gcd(mut a: usize, mut b: usize) -> usize {
    if a > b {
        (a, b) = (b, a);
    }
    while a != 0 {
        (a, b) = (b % a, a);
    }
    b
}
